### PR TITLE
Autocalibration tools

### PR DESCRIPTION
**Pull request by markbandstra**
This commit adds some basic tools to automatically calibrate a `Spectrum` using minimal information about its gain. The notebook `autocal.ipynb` explains the process, but it amounts to:

1. Create a `PeakFilter` using a rough FWHM (in channels) at one channel, and use it to instantiate a `PeakFinder`.
2. Use `PeakFinder.find_peaks` to find the highest signal-to-noise peaks in the spectrum.
3. Instantiate an `AutoCalibrator` with a `PeakFinder`.
4. Use `AutoCalibrator.fit` with a list of energies to match to the peak candidates.

Remaining TODO items:
- [x] Split peak-finding functionality out of `AutoCalibrator` into `PeakFinder`
- [x] Use `PeakFinder` to estimate both mean and FWHM of each line.
- [x] Add tests.
- [x] Update readme for `sim_spec`
- [x] Ensure usability of the API and robustness with other real-world examples.
- [x] Organization: should `autocal` remain in `core` or move to `tools`?
- ~[ ] Integration: integrate autocalibration into a `Spectrum` method?~

### CODE REVIEW

**Comment by markbandstra** [comment]
This PR is complete from my perspective.

**Comment by jccurtis** [comment]
Excellent work @markbandstra. I'll have a look tomorrow.

File: becquerel/core/autocal.py [hunk 1]
```diff
@@ -0,0 +1,267 @@
...
+        F = sum(snr^2 * (x * g - y)^2 / x)
+        dF/dg = sum(2 * snr^2 * x * (x * g - y) / x)
+              = 2 * sum(snr^2 * (x * g - y))
+              = 2 * (g * sum(snr^2 * x) - sum(snr^2 * y))
+              = 0
+        g = sum(snr^2 * y) / sum(snr^2 * x)
+
+    """
+    assert len(channels) == len(energies)
+    assert len(channels) == len(snrs)
```
Comment (jccurtis) [inline becquerel/core/autocal.py]: @markbandstra can you change these from `assert` to raise an `AutoCalibrationError`?

File: becquerel/core/autocal.py [hunk 2]
```diff
@@ -0,0 +1,267 @@
...
+    F_j = snr_j^2 (g_j x_j - y_j)^2 / x_j
+
+    Leave-one-out cross validation FOM:
+    F_LOO = sum_j(F_j) / N
+    = sum_j(snr_j^2 (g_j x_j - y_j)^2 / x_j) / N
+    = sum_j(snr_j^2 (s2y x_j - s2x y_j)^2 / (s2x - snr_j^2 x_j)^2 / x_j) / N
+
+    """
+    assert len(channels) == len(energies)
+    assert len(channels) == len(snrs)
```
Comment (jccurtis) [inline becquerel/core/autocal.py]: @markbandstra can you change these from `assert` to raise an `AutoCalibrationError`?

File: becquerel/core/autocal.py [hunk 3]
```diff
@@ -0,0 +1,267 @@
...
+
+
+def find_best_gain(
+        channels, snrs, required_energies, optional=(),
+        gain_range=(1e-3, 1e3), de_max=10., verbose=False):
+    """Find the gain that gives the best match of peaks to energies."""
+    assert len(channels) == len(snrs)
+    assert len(channels) >= 2
+    assert len(required_energies) >= 2
+    assert len(channels) >= len(required_energies)
```
Comment (jccurtis) [inline becquerel/core/autocal.py]: @markbandstra can you change these from `assert` to raise an `AutoCalibrationError`?

File: becquerel/core/autocal.py [hunk 4]
```diff
@@ -0,0 +1,267 @@
...
+        """Reset all of the members."""
+        self.gain = None
+        self.cal = None
+        self.fit_channels = []
+        self.fit_snrs = []
+        self.fit_energies = []
+
+    def set_peaks(self, peakfinder):
+        """Use the peaks found by the PeakFinder."""
+        assert isinstance(peakfinder, PeakFinder)
```
Comment (jccurtis) [inline becquerel/core/autocal.py]: @markbandstra can you change these from `assert` to raise an `AutoCalibrationError`?

File: becquerel/core/peakfinder.py [hunk 5]
```diff
@@ -0,0 +1,367 @@
...
+
+    The width of the kernel scales proportionally to the square root
+    of the channels, with a minimum value set by 'fwhm_at_0'.
+
+    """
+
+    def __init__(self, ref_channel, ref_fwhm, fwhm_at_0=1.):
+        """Initialize with a reference line position and FWHM in channels."""
+        assert ref_channel > 0
+        assert ref_fwhm > 0
```
Comment (jccurtis) [inline becquerel/core/peakfinder.py]: @markbandstra can you change these from `assert` to raise an `PeakFinderError`?

File: becquerel/core/peakfinder.py [hunk 6]
```diff
@@ -0,0 +1,367 @@
...
+        if n_center % 2 == 0:
+            n_center += 1
+        kernel0 = -0.5 * np.ones(n_center)
+        kernel0 = np.append(kernel0, np.ones(n_center))
+        kernel0 = np.append(kernel0, -0.5 * np.ones(n_center))
+        n_side = len(kernel0) // 2
+        kernel = np.zeros(2 * n_side + n_channels)
+        kernel[channel:channel + len(kernel0)] = kernel0[:]
+        kernel = kernel[n_side:-n_side]
+        assert len(kernel) == n_channels
```
Comment (jccurtis) [inline becquerel/core/peakfinder.py]: @markbandstra can you change these from `assert` to raise a `PeakFinderError`?

File: becquerel/core/peakfinder.py [hunk 7]
```diff
@@ -0,0 +1,367 @@
...
+
+    pass
+
+
+class PeakFinder(object):
+    """Find peaks in a spectrum after convolving it with a kernel."""
+
+    def __init__(self, spectrum, kernel, min_sep=5, fwhm_tol=(0.5, 1.5)):
+        """Initialize with a spectrum and kernel."""
+        assert min_sep > 0
```
Comment (jccurtis) [inline becquerel/core/peakfinder.py]: @markbandstra can you change these from `assert` to raise a `PeakFinderError`?

File: becquerel/core/peakfinder.py [hunk 8]
```diff
@@ -0,0 +1,367 @@
...
+        """Restore peak finder to pristine starting condition."""
+        self.channels = []
+        self.snrs = []
+        self.fwhms = []
+        self.integrals = []
+        self.backgrounds = []
+
+    def sort_by(self, arr):
+        """Sort peaks by the provided array."""
+        assert len(arr) == len(self.channels)
```
Comment (jccurtis) [inline becquerel/core/peakfinder.py]: @markbandstra can you change these from `assert` to raise a `PeakFinderError`?

File: becquerel/core/peakfinder.py [hunk 9]
```diff
@@ -0,0 +1,367 @@
...
+        self.channels = list(self.channels[i])
+        self.snrs = list(self.snrs[i])
+        self.fwhms = list(self.fwhms[i])
+        self.integrals = list(self.integrals[i])
+        self.backgrounds = list(self.backgrounds[i])
+
+    def calculate(self, spectrum, kernel):
+        """Calculate the convolution of the spectrum with the kernel."""
+        assert isinstance(spectrum, Spectrum)
+        assert isinstance(kernel, PeakFilter)
```
Comment (jccurtis) [inline becquerel/core/peakfinder.py]: @markbandstra can you change these from `assert` to raise a `PeakFinderError`?

File: becquerel/core/peakfinder.py [hunk 10]
```diff
@@ -0,0 +1,367 @@
...
+            # snr(chan) = snr(chan0) - 0.5 d2snr/dchan2(chan0) (chan-chan0)^2
+            # 0.5 = 1 - 0.5 d2snr/dchan2 (fwhm/2)^2 / snr0
+            # 1 = d2snr/dchan2 (fwhm/2)^2 / snr0
+            # fwhm = 2 sqrt(snr0 / d2snr/dchan2)
+            fwhm0 = self.kernel.fwhm(chan)
+            h = int(max(1, 0.2 * fwhm0))
+            d2 = (1 * self.snr[chan - h]
+                  - 2 * self.snr[chan]
+                  + 1 * self.snr[chan + h]) / h**2
+            assert d2 < 0
```
Comment (jccurtis) [inline becquerel/core/peakfinder.py]: @markbandstra can you change these from `assert` to raise a `PeakFinderError`?

File: becquerel/core/peakfinder.py [hunk 11]
```diff
@@ -0,0 +1,367 @@
...
+                'Channel range {}-{} is invalid'.format(min_chan, max_chan))
+        if min_snr < 0:
+            raise PeakFinderError(
+                'Minimum SNR {:.3f} must be > 0'.format(min_snr))
+        if self.snr.max() < min_snr:
+            raise PeakFinderError(
+                'SNR threshold is {:.3f} but maximum SNR is {:.3f}'.format(
+                    min_snr, self.snr.max()))
+        assert isinstance(max_num, int)
+        assert max_num >= 2
```
Comment (jccurtis) [inline becquerel/core/peakfinder.py]: @markbandstra can you change these from `assert` to raise a `PeakFinderError`?

File: tests/samples/INVENTORY.md [hunk 12]
```diff
@@ -0,0 +1,13 @@
+# Sample spectra for becquerel
```
Comment (jccurtis) [inline tests/samples/INVENTORY.md]: For better PR records can this be moved to another PR?

**Comment by jccurtis** [comment]
@markbandstra I'll have to look at the notebooks on a local machine since they aren't rendering on Github for some reason (looks like 404 not found error from the AWS CDN, weird). Other notebooks render alright but those are on a different branch. In any case I'll have a look in a couple of days.

File: tests/samples/INVENTORY.md (hunk 12 above)
Comment (markbandstra) [reply tests/samples/INVENTORY.md]: Sure, however it could make sense here because I am adding 3 new files to demonstrate and test the code in this PR (3 of a total of 9 files now).

Comment (jccurtis) [reply tests/samples/INVENTORY.md]: Totally, I think we should just add it quickly in another PR with the current files and then add the relevant lines in this PR. I'm interested in seeing in the milestone list so it is easier to track updates

Comment (jccurtis) [reply tests/samples/INVENTORY.md]: See #138

**Comment by cosama** [comment]
I just started to have a quick look at this too. Just a detail, but at least on my system it seems that jupyter does not automatically set `%matplotlib inline`, so I have to add either that line to the top or `plt.show()` commands whenever a plot should be created. I think the inline command would be easiest. Might be also one of the reason @jccurtis has issues seeing the notebooks on github (although for me they render just fine). 

We might want to update the plotting notebook at some point too to be consistent.

**Comment by jccurtis** [comment]
I agree with the inline magic command. 

The notebook rendering was a resource error with the actual notebook data stored on the AWS backend. That was probably a temporary glitch.

File: becquerel/core/peakfinder.py [hunk 13]
```diff
@@ -0,0 +1,395 @@
...
+    def kernel(self, channel, n_channels):
+        """Generate the kernel for the given channel."""
+        raise NotImplementedError
+
+    def kernel_matrix(self, n_channels):
+        """Build a matrix of the kernel evaluated at each channel."""
+        kern = np.zeros((n_channels, n_channels))
+        for j in range(n_channels):
+            kern[:, j] = self.kernel(j, n_channels)
+        kern_pos = +1 * kern.clip(0, np.inf)
```
Comment (chunhochow) [inline becquerel/core/peakfinder.py]: do we want to do, e.g. `kern.clip(0, np.inf)  # * (+1)`, so we don't have to do an extra unnecessary multiplication operation?

File: becquerel/core/peakfinder.py [hunk 14]
```diff
@@ -0,0 +1,395 @@
...
+            extent=[n_channels, 0, 0, n_channels])
+        plt.colorbar()
+        plt.xlabel('Input channel')
+        plt.ylabel('Output channel')
+        plt.gca().set_aspect('equal')
+
+    def convolve(self, data):
+        """Convolve this kernel with the data."""
+        kern_mat = self.kernel_matrix(len(data))
+        kern_mat_pos = +1 * kern_mat.clip(0, np.inf)
```
Comment (chunhochow) [inline becquerel/core/peakfinder.py]: do we want to do, e.g. kern_mat.clip(0, np.inf) # * (+1), so we don't have to do an extra unnecessary multiplication operation?

File: becquerel/core/peakfinder.py (hunk 13 above)
Comment (markbandstra) [reply becquerel/core/peakfinder.py]: The multiplication by +1 is for parallelism with the line below.

File: becquerel/core/peakfinder.py (hunk 14 above)
Comment (markbandstra) [reply becquerel/core/peakfinder.py]: The multiplication by +1 is for parallelism with the line below (see also line 71).

**Comment by jccurtis** [comment]
@cosama and I will run the notebook and provide more feedback
